Source/PartnerCenter.CustomerPortal/BusinessLogic/Commerce/PaymentGateways/PayUGateway.cs

```diff
@@ -0,0 +1,437 @@
+        {
+            ////PayU api not provided
+            await Task.FromResult(string.Empty);
+        }
+
+        /// <summary>
+        /// Voids an authorized payment with PayUMoney.
+        /// </summary>
+        /// <param name="authorizationCode">The authorization code for the payment to void.</param>
+        /// <returns>a Task</returns>
+        public async Task VoidAsync(string authorizationCode)
+        {
+            authorizationCode.AssertNotEmpty(nameof(authorizationCode));
+
+            // given the authorizationId string... Lookup the authorization to void it. 
+            try
+            {
+                RefundResponse refundResponse = await ApiCalls.RefundPayment(this.payerId, authorizationCode);
+                if (refundResponse.Status != 0 || !refundResponse.Message.Equals("Refund Initiated"))
+                {
+                    throw new Exception("Error in refund");
+                }
+            }
+            catch (PartnerDomainException ex)
+            {
+                this.ParsePayUException(ex);
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the order details maintained for the payment gateway.  
+        /// </summary>
+        /// <param name="payerId">The Payer Id.</param>
+        /// <param name="paymentId">The Payment Id.</param>
+        /// <param name="orderId">The Order Id.</param>
+        /// <param name="customerId">The Customer Id.</param>
+        /// <returns>The order associated with this payment transaction.</returns>
+        public async Task<OrderViewModel> GetOrderDetailsFromPaymentAsync(string payerId, string paymentId, string orderId, string customerId)
+        {
+            // this payment gateway ignores orderId & customerId. 
+            payerId.AssertNotEmpty(nameof(payerId));
+            paymentId.AssertNotEmpty(nameof(paymentId));
+
+            this.payerId = payerId;
+            this.paymentId = paymentId;
+
+            return await this.GetOrderDetails();
+        }
+
+        /// <summary>
+        /// Generate hash. 
+        /// </summary>
+        /// <param name="text">hash string.</param>
+        /// <returns>return string</returns>
+        public string Generatehash512(string text)
+        {
+            byte[] message = Encoding.UTF8.GetBytes(text);
+
+            UnicodeEncoding ue = new UnicodeEncoding();
+            byte[] hashValue;
+            System.Security.Cryptography.SHA512Managed hashString = new System.Security.Cryptography.SHA512Managed();
+            string hex = string.Empty;
+            hashValue = hashString.ComputeHash(message);
+            foreach (byte x in hashValue)
+            {
+                hex += string.Format("{0:x2}", x);
+            }
+
+            return hex;
+        }
+
+        /// <summary>
+        /// Generate transaction id. 
+        /// </summary>
+        /// <returns>return string</returns>
+        public string Generatetxnid()
```

Comment:
>   public string Generatetxnid() [](start = 5, length = 32)

method name conventions??? GenerateTransactionId() seems to be a good choice. And shouldn't this method be private?

Reply:
Done.

---
In reply to: [183242976](https://github.com/PartnerCenterSamples/Reseller-Web-Application/pull/61#discussion_r183242976) [](ancestors = 183242976)